parse_node/parse_node_with_comments.ts

```diff
@@ -0,0 +1,180 @@
+    }
+
+    break
+  }
+
+  return count
+}
+
+export const parseNodeWithComments = (
+  node: ts.Node,
+  props: ParseState
+): ParseNodeType => {
+  if (ts.isSourceFile(node)) {
+    return parseNodeWithoutComments(node, props)
+  }
+
+  let leadingComments =
+    ts
+      .getLeadingCommentRanges(
+        node.getSourceFile().getFullText(),
+        node.getFullStart()
+      )
+      ?.filter((v) => v.kind === ts.SyntaxKind.SingleLineCommentTrivia) ?? []
+
+  let tarilingComments =
+    ts
+      .getTrailingCommentRanges(
+        node.getSourceFile().getFullText(),
+        node.getFullStart() + node.getFullWidth()
+      )
+      ?.filter((v) => v.kind === ts.SyntaxKind.SingleLineCommentTrivia) ?? []
+
+  let commentStackUnwind = props.commentsStack?.length ?? 0
+
+  if (leadingComments.length > 0 || tarilingComments.length > 0) {
+    if (!props.commentsStack) {
+      props.commentsStack = []
+    }
+
+    leadingComments = leadingComments.filter(
+      (v) =>
+        !props.commentsStack!.find((c) => c.pos === v.pos && c.end === v.end)
+    )
+    tarilingComments = tarilingComments.filter(
+      (v) =>
+        !props.commentsStack!.find((c) => c.pos === v.pos && c.end === v.end)
+    )
+
+    props.commentsStack.push(...leadingComments, ...tarilingComments)
+  }
+
+  const result = parseNodeWithoutComments(node, props)
+
+  if (leadingComments.length > 0 || tarilingComments.length > 0) {
+    const fullText = node.getSourceFile().getFullText()
+
+    if (leadingComments.length > 0) {
+      const leadingCommentsText = leadingComments.map(
+        (v) => "#" + fullText.slice(v.pos + 2, v.end)
+      )
+      const addFirstNewLineToLeadingComments =
+        getBackwardsClosestNewLinesCount(fullText, leadingComments[0].pos) >
+          1 || !result.content.endsWith("\n")
+
+      result.content =
+        (addFirstNewLineToLeadingComments ? "\n" : "") +
+        leadingCommentsText.join("\n") +
+        "\n" +
+        result.content
+    }
+
+    if (tarilingComments.length > 0) {
+      const trailingCommentsText = tarilingComments.map(
+        (v) => "#" + fullText.slice(v.pos + 2, v.end)
+      )
+      const addFirstNewLineToTrailingComments =
+        getBackwardsClosestNewLinesCount(fullText, tarilingComments[0].pos) > 0
+
+      result.content =
+        result.content +
+        (addFirstNewLineToTrailingComments ? "\n" : " ") +
+        trailingCommentsText.join("\n") +
+        "\n"
+    }
+
+    props.commentsStack!.length = commentStackUnwind
+  }
+
+  return result
+}
+
+export const testComments: Test = {
+  keepComments: true,
+  ts: `
+// This is a test class
+// abc
+class Test extends Area2D {
+  // This a constructor
+  constructor() {
+    // This is a super call
+
+  }
+
+  // this is a method
+  method() {
+    // this is a print
+    print("a")
+
+    // This is an if statement
+    if (2 == 3) {
+      // This is an empty block
+    }
+
+    print(/* this is block comment */"l")
+
+    print(
```

Comment:
If you do decide to go by the per-line route, I think the proper place to inject the comments would be here:

https://github.com/johnfn/ts2gd/blob/main/parse_node.ts#L195

Basically all code in ts2gd flows through `combine` so this is a pretty safe place to put it. We have this notion of `beforeLines` and `afterLines` that allow you to say "emit this code as a standalone line before/after the line currently being codegened" so we could use that for comments as well.

Reply:
> It's extremely cool that we can do inline comments without pulling them out into the previous statement... but does it work in the even more general case? I mean, TS can handle all sorts of weird stuff, like:

I'm a bit confused because this example `print(/* this is block comment */"l")` shows that all block/multiline `/* ... */` comments are currently strpped out. Did I misunderstand something?

Reply:
Oh - good point on the block comments, I glossed over that. However, what about something like this?

```
let foo = { bar: 1 }
console.log(foo. 
  // Hello
bar)

```

Reply:
I realize that specifically may seem like an obscure case, but I think the thing that I'm more generally worried about is that there are possibly a lot of cases like this: you can really put comments almost anywhere in JS/TS, but if you map that 1:1 to Godot, there might be a lot of weird cases that don't compile correctly.